README.md
# Sisimai parser plugin for Embulk

Embulk parser plugin for [Sisimai](https://github.com/sisimai/rb-Sisimai)
bounce mail analyzing interface(A successor to bounceHammer).

## Overview

* **Plugin type**: parser
* **Guess supported**: no

## Configuration

- **format**: output format (`column`,`json` or `sisito`, default: `column`)
- **extract_mail_address**: extract mail_address into user, host and verp parts(bool, default: false).
- **include_delivered**: include delivered mail Status: 2.X.Y, (boolean, default: `false`)

The ``extract_mail_address`` parameter is `column` format mode only.
And the `include_delivered` parameter can't use format `sisito`.

## Example

### format: column

```yaml
in:
  type: any file input plugin type
  parser:
    type: sisimai
    format: column
```

Example output

``extract_mail_address``: ``false`` (default)

```
        action (   string) : failed
     addresser (   string) : user1@example.jp
         alias (   string) : 
deliverystatus (   string) : 5.1.2
   destination (   string) : example.gov
diagnosticcode (   string) : 550 Host unknown
diagnostictype (   string) : SMTP
  feedbacktype (   string) : 
         lhost (   string) : 192.0.2.97
        listid (   string) : 
     messageid (   string) : AA406E7E18714AB2927DAACC24B47C4A@USER-PC97
        reason (   string) : hostunknown
     recipient (   string) : domain-does-not-exist@example.gov
     replycode (   string) : 550
         rhost (   string) : example.gov
  senderdomain (   string) : example.jp
     smtpagent (   string) : Sendmail
   smtpcommand (   string) : 
    softbounce (     long) : 0
       subject (   string) : MULTIBYTE CHARACTERS HAVE BEEN REMOVED
     timestamp (timestamp) : 2008-09-18 08:54:04 UTC
timezoneoffset (   string) : +0900
         token (   string) : d059e55e074333fe59001b1d30d27da85a1a9c1d
```

``extract_mail_address``: ``true``

```
        action (   string) : failed
     addresser (   string) : user1@example.jp
         alias (   string) : 
deliverystatus (   string) : 5.1.2
   destination (   string) : example.gov
diagnosticcode (   string) : 550 Host unknown
diagnostictype (   string) : SMTP
  feedbacktype (   string) : 
         lhost (   string) : 192.0.2.97
        listid (   string) : 
     messageid (   string) : AA406E7E18714AB2927DAACC24B47C4A@USER-PC97
        reason (   string) : hostunknown
     recipient (   string) : domain-does-not-exist@example.gov
     replycode (   string) : 550
         rhost (   string) : example.gov
  senderdomain (   string) : example.jp
     smtpagent (   string) : Sendmail
   smtpcommand (   string) : 
    softbounce (     long) : 0
       subject (   string) : MULTIBYTE CHARACTERS HAVE BEEN REMOVED
     timestamp (timestamp) : 2008-09-18 08:54:04 UTC
timezoneoffset (   string) : +0900
         token (   string) : d059e55e074333fe59001b1d30d27da85a1a9c1d
addresser_user (   string) : user1
addresser_host (   string) : example.jp
addresser_vrep (   string) : 
recipient_user (   string) : domain-does-not-exist
recipient_host (   string) : example.gov
recipient_vrep (   string) : 
```


### format: json

```yaml
in:
  type: any file input plugin type
  parser:
    type: sisimai
    format: json
```

```
result (json) : { "token": "d059e55e074333fe59001b1d30d27da85a1a9c1d", "lhost": "192.0.2.97", "rhost": "example.gov", "listid": "", "alias": "", "reason": "hostunknown", "subject": "MULTIBYTE CHARACTERS HAVE BEEN REMOVED", "messageid": "AA406E7E18714AB2927DAACC24B47C4A@USER-PC97", "smtpagent": "Sendmail", "smtpcommand": "", "destination": "example.gov", "diagnosticcode": "550 Host unknown", "senderdomain": "example.jp", "deliverystatus": "5.1.2", "timezoneoffset": "+0900", "feedbacktype": "", "diagnostictype": "SMTP", "action": "failed", "replycode": "550", "softbounce": 0, "addresser": "user1@example.jp", "recipient": "domain-does-not-exist@example.gov", "timestamp": 1221728044 }
```


```json
{
  "token": "d059e55e074333fe59001b1d30d27da85a1a9c1d",
  "lhost": "192.0.2.97",
  "rhost": "example.gov",
  "listid": "",
  "alias": "",
  "reason": "hostunknown",
  "subject": "MULTIBYTE CHARACTERS HAVE BEEN REMOVED",
  "messageid": "AA406E7E18714AB2927DAACC24B47C4A@USER-PC97",
  "smtpagent": "Sendmail",
  "smtpcommand": "",
  "destination": "example.gov",
  "diagnosticcode": "550 Host unknown",
  "senderdomain": "example.jp",
  "deliverystatus": "5.1.2",
  "timezoneoffset": "+0900",
  "feedbacktype": "",
  "diagnostictype": "SMTP",
  "action": "failed",
  "replycode": "550",
  "softbounce": 0,
  "addresser": "user1@example.jp",
  "recipient": "domain-does-not-exist@example.gov",
  "timestamp": 1221728044
}
```

### format: sisito

[sisito](https://github.com/winebarrel/sisito) is the GUI interface for libsisimai.
The following cofiguration import bounce mails into sisito database.

```yaml
in:
  path_prefix: path/to/maildir
  type: file
  parser:
    type: sisimai
    format: sisito
out:
  type: mysql
  host: localhost
  user: user
  password: password
  mode: replace
  database: maillog_test
  table: bounce_mails # Don't change. sisito use this table name
  column_options:
    created_at: { type: datetime }
    updated_at: { type: datetime }
```


## Install

```
$ embulk gem install embulk-parser-sisimai
```

## Build

```
$ rake
```


CHANGES.md
0.1.1 2017-09-20
----------------

* NEW sisito format support

0.1.0 2016-02-18
----------------

* First release


lib/embulk/parser/sisimai_analyzer.rb
require 'sisimai'
require 'sisimai/message'
require 'sisimai/data'
require 'pp'

module Embulk
  module Parser

    class SisimaiAnalyzer < ParserPlugin
      Plugin.register_parser("sisimai_analyzer", self)

      def self.transaction(config, &control)
        task = {
          "format" => config.param("format", :string, default: "column"),
          "extract_mail_address" => config.param("extract_mail_address", :bool, default: false)
        }

        format = task["format"]
        columns = case format
        when "json"
          [ Column.new(0, "result", :json) ]
        when "column"
          c = [
            Column.new(0, "action", :string),
            Column.new(1, "addresser",:string),
            Column.new(2, "alias", :string),
            Column.new(3, "deliverystatus", :string),
            Column.new(4, "destination", :string),
            Column.new(5, "diagnosticcode", :string),
            Column.new(6, "diagnostictype", :string),
            Column.new(7, "feedbacktype", :string),
            Column.new(8, "lhost", :string),
            Column.new(9, "listid", :string),
            Column.new(10, "messageid", :string),
            Column.new(11, "reason", :string),
            Column.new(12, "recipient", :string),
            Column.new(13, "replycode", :string),
            Column.new(14, "rhost", :string),
            Column.new(15, "senderdomain", :string),
            Column.new(16, "smtpagent", :string),
            Column.new(17, "smtpcommand", :string),
            Column.new(18, "softbounce", :long),
            Column.new(19, "subject", :string),
            Column.new(20, "timestamp", :timestamp),
            Column.new(21, "timezoneoffset", :string),
            Column.new(22, "token", :string),
          ]
          if task['extract_mail_address'] == true
            c += [
              Column.new(23, "addresser_user", :string),
              Column.new(24, "addresser_host", :string),
              Column.new(25, "addresser_vrep", :string),
              Column.new(26, "recipient_user", :string),
              Column.new(27, "recipient_host", :string),
              Column.new(28, "recipient_vrep", :string),
            ]
          end
          c
        else
          raise ArgumentError,"Unkown format type: #{format}"
        end

        yield(task, columns)
      end

      def init
        # initialization code:
        @format = task["format"]
        @extract_mail_address = task["extract_mail_address"]
      end

      def run(file_input)
        while file = file_input.next_file
          mesg = Sisimai::Message.new( data: file.read )
          datas = Sisimai::Data.make( data: mesg )
          if datas.nil?
            Embulk.logger.info "This file does not contaion bounce mail. skip."
            next
          end
          datas.each do |data|
            case @format
            when "json"
              page_builder.add([ data.dump ])
            when "column"
              column_data = make_column_array(data)

              page_builder.add(column_data)
            else
              raise RuntimeError,"Invalid format #{@format}"
            end
          end
        end
        page_builder.finish
      end
      private
      def make_column_array(data)
        row = [
          data.action,
          data.addresser.to_json,
          data.alias,
          data.deliverystatus,
          data.destination,
          data.diagnosticcode,
          data.diagnostictype,
          data.feedbacktype,
          data.lhost,
          data.listid,
          data.messageid,
          data.reason,
          data.recipient.to_json,
          data.replycode,
          data.rhost,
          data.senderdomain,
          data.smtpagent,
          data.smtpcommand,
          data.softbounce,
          data.subject,
          data.timestamp.to_time.utc,
          data.timezoneoffset,
          data.token,
        ]
        if @extract_mail_address
          row += [
            data.addresser.user,
            data.addresser.host,
            data.addresser.verp,
            data.recipient.user,
            data.recipient.host,
            data.recipient.verp,
          ]
        end
        row
      end
    end
  end
end


lib/embulk/parser/sisimai.rb
require 'sisimai'
require 'sisimai/message'
require 'sisimai/data'
require 'digest/sha1'

module Embulk
  module Parser

    class Sisimai < ParserPlugin
      Plugin.register_parser("sisimai", self)

      def self.transaction(config, &control)
        task = {
          "format" => config.param("format", :string, default: "column"),
          "extract_mail_address" => config.param("extract_mail_address", :bool, default: false),
          "include_delivered" => config.param("include_delivered", :bool, default: false)
        }

        format = task["format"]
        columns = case format
        when "json"
          [ Column.new(0, "result", :json) ]
        when "column"
          c = [
            Column.new(0, "action", :string),
            Column.new(1, "addresser",:string),
            Column.new(2, "alias", :string),
            Column.new(3, "deliverystatus", :string),
            Column.new(4, "destination", :string),
            Column.new(5, "diagnosticcode", :string),
            Column.new(6, "diagnostictype", :string),
            Column.new(7, "feedbacktype", :string),
            Column.new(8, "lhost", :string),
            Column.new(9, "listid", :string),
            Column.new(10, "messageid", :string),
            Column.new(11, "reason", :string),
            Column.new(12, "recipient", :string),
            Column.new(13, "replycode", :string),
            Column.new(14, "rhost", :string),
            Column.new(15, "senderdomain", :string),
            Column.new(16, "smtpagent", :string),
            Column.new(17, "smtpcommand", :string),
            Column.new(18, "softbounce", :long),
            Column.new(19, "subject", :string),
            Column.new(20, "timestamp", :timestamp),
            Column.new(21, "timezoneoffset", :string),
            Column.new(22, "token", :string),
          ]
          if task['extract_mail_address'] == true
            c += [
              Column.new(23, "addresser_user", :string),
              Column.new(24, "addresser_host", :string),
              Column.new(25, "addresser_vrep", :string),
              Column.new(26, "recipient_user", :string),
              Column.new(27, "recipient_host", :string),
              Column.new(28, "recipient_vrep", :string),
            ]
          end
          c
        when "sisito"
          c = [
            Column.new(0, "timestamp", :timestamp),
            Column.new(1, "lhost", :string),
            Column.new(2, "rhost", :string),
            Column.new(3, "alias", :string),
            Column.new(4, "listid", :string),
            Column.new(5, "reason", :string),
            Column.new(6, "action", :string),
            Column.new(7, "subject", :string),
            Column.new(8, "messageid", :string),
            Column.new(9, "smtpagent", :string),
            Column.new(10, "softbounce", :long),
            Column.new(11, "smtpcommand", :string),
            Column.new(12, "destination", :string),
            Column.new(13, "senderdomain", :string),
            Column.new(14, "feedbacktype", :string),
            Column.new(15, "diagnostictype", :string),
            Column.new(16, "deliverystatus", :string),
            Column.new(17, "timezoneoffset", :string),
            Column.new(18, "addresser",:string),
            Column.new(19, "recipient", :string),
            Column.new(20, "addresseralias",:string),
            Column.new(21, "digest",:string),
            Column.new(22, "created_at",:timestamp),
            Column.new(23, "updated_at",:timestamp),
          ]
        else
          raise ArgumentError,"Unkown format type: #{format}"
        end

        format = task["format"]
        inc_delivered = task["include_delivered"]
        extract_mail_address = task["extract_mail_address"]
        Embulk.logger.info "sisimai format: #{format} include_delivered: #{inc_delivered}, extract_mail_address: #{extract_mail_address}"
        yield(task, columns)
      end

      def init
        # initialization code:
        @format = task["format"]
        @inc_delivered = task["include_delivered"]
        @extract_mail_address = task["extract_mail_address"]
      end

      def run(file_input)
        while file = file_input.next_file
          begin
            # Sisimai expects input data is UTF-8 string.
            src = file.read.force_encoding(Encoding::UTF_8)
            mesg = ::Sisimai::Message.new( data: src )
            datas = ::Sisimai::Data.make( data: mesg, delivered: @inc_delivered )
          rescue
            Embulk.logger.error "Error #{$!} #{src}"
            raise
          end
          if datas.nil?
            Embulk.logger.info "This file does not contaion bounce mail. skip."
            next
          end
          datas.each do |data|
            case @format
            when "json"
              page_builder.add([ data.dump ])
            when "column"
              column_data = make_column_array(data)

              page_builder.add(column_data)
            when "sisito"
              column_data = make_sisito_array(data)
              page_builder.add(column_data)
            else
              raise RuntimeError,"Invalid format #{@format}"
            end
          end
        end
        page_builder.finish
      end
      private
      def make_column_array(data)
        row = [
          data.action,
          data.addresser.to_json,
          data.alias,
          data.deliverystatus,
          data.destination,
          data.diagnosticcode,
          data.diagnostictype,
          data.feedbacktype,
          data.lhost,
          data.listid,
          data.messageid,
          data.reason,
          data.recipient.to_json,
          data.replycode,
          data.rhost,
          data.senderdomain,
          data.smtpagent,
          data.smtpcommand,
          data.softbounce,
          data.subject,
          data.timestamp.to_time.utc,
          data.timezoneoffset,
          data.token,
        ]
        if @extract_mail_address
          row += [
            data.addresser.user,
            data.addresser.host,
            data.addresser.verp,
            data.recipient.user,
            data.recipient.host,
            data.recipient.verp,
          ]
        end
        row
      end

      def make_sisito_array(data)
        #data.diagnosticcode,
        #data.replycode,
        #data.token,

        addresseralias = data.addresser.alias
        addresseralias = data.addresser.to_s if addresseralias.empty?

        now = Time.now
        row = [
          data.timestamp.to_time.utc,
          data.lhost,
          data.rhost,
          data.alias,
          data.listid,
          data.reason,
          data.action,
          data.subject,
          data.messageid,
          data.smtpagent,
          data.softbounce,
          data.smtpcommand,
          data.destination,
          data.senderdomain,
          data.feedbacktype,
          "", # data.diagnostictype,
          data.deliverystatus,
          data.timezoneoffset,
          data.addresser.to_json,
          data.recipient.to_json,
          addresseralias,
          Digest::SHA1.hexdigest(data.recipient.to_s),
          now,
          now
        ]
      end
    end
  end
end
